### PR TITLE
feat: Internal column support

### PR DESCRIPTION
**Pull request by lichuang**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

feat: Internal column support.

1. add more block and segment meta data into`BlockMetaIndex` in order to generate internal column values.
2. introduce `FillInternalColumnProcessor` to fill internal columns data.
3. unit test(see `05_0031_internal_column`)

Closes https://github.com/datafuselabs/databend/issues/10081

### CODE REVIEW

File: src/query/expression/src/schema.rs [hunk 1]
```diff
@@ -382,6 +382,17 @@ impl TableSchema {
         Ok(())
     }
 
+    // Every internal column has constant column id, no need to generate column id of internal columns.
```
[the remark this answers is not in the document]
Comment (lichuang) [reply src/query/expression/src/schema.rs:385]: it means that internal column has constant column id, not generated like normal column.

File: src/query/sql/src/planner/binder/internal_column_factory.rs [hunk 2]
```diff
@@ -23,10 +23,9 @@ use common_catalog::plan::ROW_ID;
 use common_catalog::plan::SEGMENT_NAME;
 use common_catalog::plan::SNAPSHOT_NAME;
 use common_exception::Result;
-use common_expression::types::DataType;
 
 pub struct InternalColumnFactory {
-    internal_columns: BTreeMap<String, InternalColumn>,
+    pub internal_columns: BTreeMap<String, InternalColumn>,
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/internal_column_factory.rs:28]: use `#[ctor]` is enough, see `common_functions::scalars::BUILTIN_FUNCTIONS;`

File: src/query/sql/src/planner/metadata.rs [hunk 3]
```diff
@@ -304,20 +344,31 @@ pub struct DerivedColumn {
...
+
 #[derive(Clone, Debug)]
 pub enum ColumnEntry {
     /// Column from base table, for example `SELECT t.a, t.b FROM t`.
     BaseTableColumn(BaseTableColumn),
 
     /// Column synthesized from other columns, for example `SELECT t.a + t.b AS a FROM t`.
     DerivedColumn(DerivedColumn),
+
+    /// Virtual columns, such as `_row_id`, `_segment_name`, etc.
```
Comment (b41sh) [inline src/query/sql/src/planner/metadata.rs:362]: Internal columns

**Review by b41sh — APPROVED** [review]

**Comment by lichuang** [comment]
@sundy-li @dantengsky

File: src/query/catalog/src/plan/internal_column.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            InternalColumnType::SnapshotName => u32::MAX - 3,
+        }
+    }
+
+    pub fn generate_column_values(&self, meta: &InternalColumnMeta, num_rows: usize) -> BlockEntry {
+        match &self.column_type {
+            InternalColumnType::RowId => {
+                let block_id = meta.block_id as u64;
+                let seg_id = meta.segment_id as u64;
+                let high_32bit = (seg_id << 48) + (block_id << 32);
```
Comment (sundy-li) [inline src/query/catalog/src/plan/internal_column.rs:99]: Num of Segment may be larger than `1<<16` (65536).

How about making `row_id =  { block_id  << 32 }    +  { block_row_id}`

File: tests/sqllogictests/suites/base/05_ddl/05_0031_internal_column [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+CREATE TABLE `05_0031_t_1`(_row_id int)
+
+statement ok
+INSERT INTO TABLE `05_0031_t` values(1),(2)
+
+statement ok
+INSERT INTO TABLE `05_0031_t` values(3)
+
+query II
+SELECT a,_row_id FROM `05_0031_t` order by _row_id
```
Comment (sundy-li) [inline tests/sqllogictests/suites/base/05_ddl/05_0031_internal_column:23]: 1. Need more tests to cover other virtual columns.
2. Need more tests to cover  more segments and blocks ( you can insert with numbers  table multiple times to simulate)

**Review by sundy-li — APPROVED** [review]

**Pull request merged**